src/ctl/ctl.c

```diff
@@ -0,0 +1,593 @@
+
+    if (n == NULL && ctl) {
+        ctl_delete_indexes(indexes);
+        indexes = NULL;
+        n = ctl_find_node(ctl->root, name, indexes);
+    }
+
+    if (n == NULL || n->type != CTL_NODE_LEAF || n->cb[type] == NULL) {
+        errno = EINVAL;
+        goto out;
+    }
+
+    ret = ctl_exec_query[type](ctx, n, source, arg, indexes);
+
+out:
+    ctl_delete_indexes(indexes);
+
+    return ret;
+}
+
+/*
+ * ctl_register_module_node -- adds a new node to the CTL tree root.
+ */
+void ctl_register_module_node(struct ctl *c, const char *name,
+                              struct ctl_node *n) {
+    struct ctl_node *nnode = c == NULL
+                                 ? &CTL_NODE(global, )[ctl_global_first_free++]
+                                 : &c->root[c->first_free++];
+
+    nnode->children = n;
+    nnode->type = CTL_NODE_NAMED;
+    nnode->name = name;
+}
+
+/*
+ * ctl_parse_query -- (internal) splits an entire query string
+ *	into name and value
+ */
+static int ctl_parse_query(char *qbuf, char **name, char **value) {
+    if (qbuf == NULL) {
+        return -1;
+    }
+
+    char *sptr = NULL;
+    *name = strtok_r(qbuf, CTL_NAME_VALUE_SEPARATOR, &sptr);
+    if (*name == NULL) {
+        return -1;
+    }
+
+    *value = strtok_r(NULL, CTL_NAME_VALUE_SEPARATOR, &sptr);
+    if (*value == NULL) {
+        return -1;
+    }
+
+    /* the value itself mustn't include CTL_NAME_VALUE_SEPARATOR */
+    char *extra = strtok_r(NULL, CTL_NAME_VALUE_SEPARATOR, &sptr);
+    if (extra != NULL) {
+        return -1;
+    }
+
+    return 0;
+}
+
+/*
+ * ctl_load_config -- executes the entire query collection from a provider
+ */
+static int ctl_load_config(struct ctl *ctl, void *ctx, char *buf) {
+    int r = 0;
+    char *sptr = NULL; /* for internal use of strtok */
+    char *name;
+    char *value;
+    char *qbuf = strtok_r(buf, CTL_STRING_QUERY_SEPARATOR, &sptr);
+
+    while (qbuf != NULL) {
+        r = ctl_parse_query(qbuf, &name, &value);
+        if (r != 0) {
+            return -1;
+        }
+
+        r = ctl_query(ctl, ctx, CTL_QUERY_CONFIG_INPUT, name, CTL_QUERY_WRITE,
+                      value);
+
+        if (r < 0 && ctx != NULL) {
+            return -1;
+        }
+
+        qbuf = strtok_r(NULL, CTL_STRING_QUERY_SEPARATOR, &sptr);
+    }
+
+    return 0;
+}
+
+/*
+ * ctl_load_config_from_string -- loads obj configuration from string
+ */
+int ctl_load_config_from_string(struct ctl *ctl, void *ctx,
+                                const char *cfg_string) {
+    char *buf = Strdup(cfg_string);
+    if (buf == NULL) {
+        return -1;
+    }
+
+    int ret = ctl_load_config(ctl, ctx, buf);
+
+    Free(buf);
+    return ret;
+}
+
+/*
+ * ctl_load_config_from_file -- loads obj configuration from file
+ *
+ * This function opens up the config file, allocates a buffer of size equal to
+ * the size of the file, reads its content and sanitizes it for ctl_load_config.
+ */
+#ifdef WINDOWS_API_NEEDED
```

Comment:
Unknown define. The code looks like it should work on linux.

Reply:
Intentionally, it's disabled due to lack of support for Windows API

Reply:
@PatKamin it is OK to not have this functionality right now - it could be added in separate PR later